src/Npgsql/TypeHandlers/NumericHandlers/NumericHandler.cs

```diff
@@ -97,7 +97,13 @@ public override decimal Read(NpgsqlReadBuffer buf, int len, FieldDescription fie
             if (sign == 0xC000)
                 throw new NpgsqlSafeReadException(new InvalidCastException("Numeric NaN not supported by System.Decimal"));
 
-            result = Math.Round(result, dscale);
+            var rscale = ((decimal.GetBits(result)[3] >> 16) & 0x7F);                           //Find scale of result
```

Comment:
`decimal.GetBits()` always allocates an `int[]`, which is really not good.

Reply:
Do you have a suggestion on a better way to do it? See my other comments (coming soon) on why I believe trailing zeros aren't just a cosmetic issue.

Reply:
@YohDeadfall I assume you're recommending a struct for `result` storage because it will be stored on the stack instead of the heap. This allows us to read the result scale which is then stored in `rscale`, also on the stack as a simple integer value.

Since popping the stack is much faster than garbage collection, we avoid a lot of overhead.

Did I get it right?

P.S. Most of my work has not needed to be so performance critical so I'm still learning this stuff.

Reply:
Yep, you're right, except one thing. Overhead is not in popping, but in memory allocation on the heap. `GetBits` allocates an array of four integers.

Don't worry about that. I'll help you with perf optimizations.

Reply:
Shouldn't the `FieldOffset` for `Mid` and `Lo` be
```
    [FieldOffset(8)]
    internal int Mid;
    [FieldOffset(12)]
    internal int Lo;
```
Instead of
```
    [FieldOffset(8)]
    internal int Lo;
    [FieldOffset(12)]
    internal int Mid;
```

Reply:
No, the order is defined in coreclr ([source](https://github.com/dotnet/coreclr/blob/85374ceaed177f71472cc4c23c69daf7402e5048/src/System.Private.CoreLib/src/System/Decimal.cs#L135-L141)).